Sources/LeafKit/LeafSyntax/LeafSyntax.swift

```diff
@@ -68,28 +172,110 @@ extension Syntax {
                 }
             }
+        }
+        
+        internal init(key: String, exports: [String : Syntax.Export], externalsSet: Set<String>, importSet: Set<String>) {
+            self.key = key
             self.exports = exports
+            self.externalsSet = externalsSet
+            self.importSet = importSet
+        }
+        
+        func externals() -> Set<String> {
+            return externalsSet
+        }
+        func imports() -> Set<String> {
+            return importSet
+        }
+        
+        func inlineRefs(_ externals: [String: LeafAST], _ imports: [String : Syntax.Export]) -> [Syntax] {
+            var newExports = [String: Export]()
+            var newImports = imports
+            var newExternalsSet = Set<String>()
+            var newImportSet = Set<String>()
+            
+            // In the case where #exports themselves contain #extends or #imports, rebuild those
+            for (key, value) in exports {
+                guard !value.externals().isEmpty || !value.imports().isEmpty else {
+                    newExports[key] = value
+                    continue
+                }
+                guard case .export(let e) = value.inlineRefs(externals, imports).first else { fatalError() }
+                newExports[key] = e
+                newExternalsSet.formUnion(e.externals())
+                newImportSet.formUnion(e.imports())
+            }
+            
+            // Now add this extend's exports onto the passed imports
+            newExports.forEach {
+                newImports[$0.key] = $0.value
+            }
+            
+            var results = [Syntax]()
+            
+            // Either return a rebuilt #extend or an inlined and (potentially partially) resolved extended syntax
+            if !externals.keys.contains(self.key) {
+                let resolvedExtend = Syntax.Extend(key: self.key,
+                                                   exports: newExports,
+                                                   externalsSet: externalsSet,
+                                                   importSet: newImportSet)
+                results.append(.extend(resolvedExtend))
+            } else {
+                // Get the external AST
+                let newAst = externals[self.key]!
+                // Remove this AST from the externals to avoid needless checks
+                let externals = externals.filter { $0.key != self.key }
+                newAst.ast.forEach {
+                    // Add each external syntax, resolving with the current available
+                    // exports and passing this extend's exports to the syntax's imports
+                    
+                    results += $0.inlineRefs(externals, newImports)
+                    // expressions may have been created by imports, convert
+                    // single parameter static values to .raw
+                    if case .expression(let e) = results.last {
```

Comment:
The nested ifs could be combined: `if case .expression(let e) = results.last, let raw = e.atomicRaw() { ... }`